File: problems/150/solution.go
package main
import "strconv"

func evalRPN(tokens []string) int {
    stack := []int{}
    
    for _, token := range tokens {
        switch token {
        case "+", "-", "*", "/":
            a, b, c := stack[len(stack) - 2], stack[len(stack) - 1], 0
            stack = stack[0: len(stack) - 2]
            
            if token == "+" {
                c = a + b
            } else if token == "*" {
                c = a * b
            } else if token == "-" {
                c = a - b
            } else {
                c = a / b
            }
            
            stack = append(stack, c)
            
        default:
            val, _ := strconv.Atoi(token)
            
            stack = append(stack, val)
        }
    }
    
    return stack[0]
}

File: problems/150/solution.cpp
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> st;
        unordered_map<string, function<int(int, int)>> ops = {
            {"+", [](const int& lhs, const int& rhs) { return lhs + rhs; }},
            {"*", [](const int& lhs, const int& rhs) { return lhs * rhs; }},
            {"/", [](const int& lhs, const int& rhs) { return lhs / rhs; }},
            {"-", [](const int& lhs, const int& rhs) { return lhs - rhs; }}
        };
        
        for (const string& token: tokens) {
            if (not ops.count(token)) {
                st.push(stoi(token));
            } else {
                int rhs = st.top();
                st.pop();
                int lhs = st.top();
                st.pop();
                
                st.push(ops[token](lhs, rhs));
            }
        }
        
        return st.top();
    }
    
    int evalRPN1(vector<string>& tokens) {
        if (tokens.empty()) {
            return 0;
        }
        
        string last = tokens.back();
        tokens.pop_back();
        
        if (last != "*" and last != "/" and last != "-" and last != "+") {
            return stoi(last);
        }
        
        int rhs = evalRPN(tokens),
            lhs = evalRPN(tokens);
        
        if (last == "*") {
            return lhs * rhs;
        } else if (last == "/") {
            return lhs / rhs;
        } else if (last == "-") {
            return lhs - rhs;
        }
        
        return lhs + rhs;
    }
};